src/features/ProfilePage/components/PublicInfo.tsx

```diff
@@ -1,35 +1,81 @@
 import styled from 'styled-components';
-import { useEffect, useState } from 'react';
+import { useState } from 'react';
 import { useTranslation } from 'react-i18next';
 
-import { selectUserInfo } from '@/features/Authentication/userSlice';
+import { selectMentor, selectUser } from '@/features/Authentication/userSlice';
 import { useAppSelector } from '@/store';
-
-import { DEFAULT_ICON_SIZE, palette } from '@/components/constants';
+import { useDebounce } from '@/hooks/useDebounce';
+import { useUpdateMentorMutation } from '@/features/MentorPage/mentorPageApi';
+import { useUpdateUserMutation } from '@/features/Authentication/authenticationApi';
+
+import {
+  DEFAULT_ICON_SIZE,
+  palette,
+  SAVE_DELAY_MS,
+} from '@/components/constants';
 import LabeledInput from '@/components/LabeledInput';
 import Slider from '@/components/Slider';
 import Text from '@/components/Text';
 import TextInput from '@/components/TextInput';
 
+import type {
+  MentorUser,
+  User,
+} from '@/features/Authentication/authenticationApi';
+
 const PublicInfo = () => {
   const { t } = useTranslation('profile');
-  const userInfo = useAppSelector(selectUserInfo);
+  const user = useAppSelector(selectUser);
+  const mentor = useAppSelector(selectMentor);
+  const [updateUser] = useUpdateUserMutation();
+  const [updateMentor] = useUpdateMentorMutation();
+
+  const saveMentorData = useDebounce(
+    (mentor: MentorUser) => updateMentor(mentor),
+    SAVE_DELAY_MS,
+  );
+
+  const saveUserAndMentorData = useDebounce(
+    (mentor: MentorUser, user: User) => {
+      updateUser(user);
+      updateMentor(mentor);
+    },
+    SAVE_DELAY_MS,
+  );
+
+  const [localData, setLocalData] = useState<MentorUser>(mentor);
+  const [skillSearchValue, setSkillSearchValue] = useState('');
+
+  const updateMentorData = (key: keyof MentorUser, value: string) => {
+    setLocalData({ ...localData, [key]: value });
+    saveMentorData({ ...mentor, [key]: value });
+  };
+
```

Comment:
Instead of this we can

```typescript
  const updateMentorData = <K extends keyof MentorUser>(
    key: K,
    value: MentorUser[K],
  ) => {
    setLocalData({ ...localData, [key]: value });
    saveMentorData({ ...mentor, [key]: value });
  };
  
  // usage, this is string
  <LabeledInput
    label={t('public.mentor.displayName')}
    onChange={(displayName) => updateMentorData('display_name', displayName)}
    value={localData.display_name}
  />
  // usage, this is number value
  <LabeledInput
    label={t('public.mentor.birthYear')}
    onChange={(birthYear) => updateMentorData('birth_year', Number(birthYear))}
    value={String(localData.birth_year)}
    type="number"
  />
  ```

Reply:
Updating the display name makes **two API calls** as opposed to the others so it needs a separate function.

Reply:
But the other ones I can combine. I just think that's a bit harder to read. But it's more concise, I'll give you that. So I'll do it. 👍